docs/training/apis-howto/api-core-ug.rst

```diff
@@ -1,212 +1,302 @@
-trial ID in the checkpoint and use it to distinguish the two types of continues.
+and ``core_context.train.report_validation_metrics()``:
 
-#. Create a new ``2_checkpoints.py`` training script called by copying the ``1_metrics.py`` script
-   from :ref:`core-metrics`.
+.. literalinclude:: ../../../examples/tutorials/core_api_pytorch_mnist/model_def_metrics.py
+   :language: python
+   :start-after: # Docs snippet start: report validation metrics
+   :end-before: # Docs snippet end: report validation metrics
+   :dedent:
 
-#. Write save and load methods for your model:
+.. note::
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/2_checkpoints.py
-      :language: python
-      :pyobject: save_state
+   Viewing Epoch-Based Metrics in the WebUI
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/2_checkpoints.py
-      :language: python
-      :pyobject: load_state
+   You can modify your code to report epoch-based metrics as shown in this example. When you report
+   epoch-based metrics during training, you'll be able to view epoch-based metric data in the WebUI.
 
-#. In your ``if __name__ == "__main__"`` block, use the ClusterInfo API to gather additional
-   information about the task running on the cluster, specifically a checkpoint to load from and the
-   trial ID, which you also pass to ``main()``.
+.. tip::
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/2_checkpoints.py
-      :language: python
-      :start-at: info = det.get_cluster_info()
+   Avoiding a Duplicate Key Error
 
-   It is recommended that you always follow this pattern of extracting values from the ClusterInfo
-   API and passing the values to lower layers of your code, instead of accessing the ClusterInfo API
-   directly in the lower layers. In this way the lower layer can be written to run on or off of the
-   Determined cluster.
+   It is best to stitch metrics together in one continuous graph as your experiment progresses. To
+   accomplish this, calculate the index used for ``steps_completed`` appropriately. The appropriate
+   calculation avoids re-writing metrics with the same index each time a new epoch begins and avoids
+   a duplicate key error.
 
-#. Within ``main()``, add logic to continue from a checkpoint, when a checkpoint is provided:
+After an epoch value has been reported, **Epoch** will be an available option for the X-Axis when
+viewing the metric data graph in the WebUI.
 
-   .. literalinclude:: ../../../examples/tutorials/core_api/2_checkpoints.py
-      :language: python
-      :start-at: def main
-      :end-at: for batch in range(starting_batch, 100)
+Step 2.3: Modify the Test Method
+================================
 
-#. You can checkpoint your model as frequently as you like. For this exercise, save a checkpoint
-   after each training report, and check for a preemption signal after each checkpoint:
+Modify the ``test()`` function header to include ``args`` and other elements you’ll need during the
```

Comment:
What are `args` about? Do they relate to `core_context` in some way? Why am I doing any of this? Think the reader would benefit from more context and forward looking explanation.

Reply:
updated

Modify the test() function header to include args and other elements you’ll need during the evaluation loop. The args variable lets us pass configuration settings such as batch size and learning rate. In addition, pass the newly created core_context into both train() and test(). Passing core_context enables reporting of metrics to the Determined master.